### PR TITLE
Enable IPC tests for Level Zero provider

### PR DESCRIPTION
**Pull request by vinser52**
### Description
Enable IPC tests for the Level Zero provider.
The main challenge was to unify work with host and device memory.

### Checklist

- [x] Code compiles without errors locally
- [x] All tests pass locally
- [x] CI workflows execute properly
- [x] New tests added, especially if they will fail without my changes

### CODE REVIEW

File: test/providers/provider_level_zero.cpp [hunk 1]
```diff
@@ -326,21 +326,66 @@ struct umfLevelZeroProviderTest : umf_test::test {
...
+    void SetUp() override {
+        test::SetUp();
+
+        auto [l0_params, accessor] = this->GetParam();
+        params = l0_params;
+        hDevice = (ze_device_handle_t)params.level_zero_device_handle;
+        hContext = (ze_context_handle_t)params.level_zero_context_handle;
+
+        if (nullptr == hDevice) {
+            GTEST_SKIP() << "Test skipped, no GPU devices found";
```
Comment (lukaszstolarczuk) [inline test/providers/provider_level_zero.cpp:350]: I think we may fail the test actually - we have a specific CMake flag to enable these tests (`UMF_BUILD_GPU_TESTS`, OFF by default). I believe we prefer not to have a "silent skips" (gtest doesn't say a word about skips).

File: test/providers/provider_level_zero.cpp [hunk 2]
```diff
@@ -58,178 +61,175 @@ struct libze_ops {
...
+        }
     }
+};
 
-    void DestroyLevelZeroOps() {
-        // Just close the handle here
-        dlclose(zeDlHandle);
-    }
+std::unique_ptr<void, DlHandleCloser> zeDlHandle = nullptr;
+//void *zeDlHandle = nullptr;
```
Comment (lukaszstolarczuk) [inline test/providers/provider_level_zero.cpp:74]: I guess you can remove this line

File: test/providers/provider_level_zero.cpp [hunk 3]
```diff
@@ -5,9 +5,12 @@
 #include <dlfcn.h>
 #include <level_zero/ze_api.h>
 
+#include "ipcFixtures.hpp"
 #include "pool.hpp"
 #include "umf/providers/provider_level_zero.h"
 
+#include <mutex>
```
Comment (lukaszstolarczuk) [inline test/providers/provider_level_zero.cpp:12]: is it needed...?

File: test/providers/provider_level_zero.cpp [hunk 4]
```diff
@@ -326,21 +326,66 @@ struct umfLevelZeroProviderTest : umf_test::test {
...
+            zeMemoryTypeExpected = ZE_MEMORY_TYPE_DEVICE;
+            break;
+        case UMF_MEMORY_TYPE_SHARED:
+            zeMemoryTypeExpected = ZE_MEMORY_TYPE_SHARED;
+            break;
+        case UMF_MEMORY_TYPE_HOST:
+            zeMemoryTypeExpected = ZE_MEMORY_TYPE_HOST;
+            break;
+        default:
+            GTEST_SKIP() << "Test skipped, unsupported memory type";
```
Comment (lukaszstolarczuk) [inline test/providers/provider_level_zero.cpp:364]: fail here as well...?

**Review by kswiecicki — APPROVED** [review]

File: test/providers/provider_level_zero.cpp (hunk 3 above)
Comment (vinser52) [reply test/providers/provider_level_zero.cpp:12]: The `InitLevelZero` function uses `std::call_once` which requires this header.

File: test/providers/provider_level_zero.cpp (hunk 2 above)
Comment (vinser52) [reply test/providers/provider_level_zero.cpp:74]: Done

File: test/providers/provider_level_zero.cpp (hunk 1 above)
Comment (vinser52) [reply test/providers/provider_level_zero.cpp:350]: Done

File: test/providers/provider_level_zero.cpp (hunk 4 above)
Comment (vinser52) [reply test/providers/provider_level_zero.cpp:364]: Done. I have removed `default` case and added `ASSERT_NE(zeMemoryTypeExpected, ZE_MEMORY_TYPE_UNKNOWN);` after `switch` block.

**Review by lukaszstolarczuk — APPROVED** [review]
thx!

**Review by bratpiorka — CHANGES_REQUESTED** [review]

File: test/ipcFixtures.hpp [hunk 5]
```diff
@@ -19,10 +19,36 @@
...
+
+class HostMemoryAccessor : public MemoryAccessor {
+  public:
+    void fill(void *ptr, size_t size, const void *pattern,
+              size_t pattern_size) override {
+        while (size) {
+            size_t copy_size = std::min(size, pattern_size);
+            std::memcpy(ptr, pattern, copy_size);
+            ptr = static_cast<char *>(ptr) + copy_size;
+            size -= copy_size;
```
[the remark this answers is not in the document]
Comment (vinser52) [reply test/ipcFixtures.hpp:37]: Strictly speaking, it should be `size >= copy_size`. And on L34 we choose minimum between `size` and `pattern_size`, so `copy_size` cannot be greater than `size`.

File: test/ipcFixtures.hpp [hunk 6]
```diff
@@ -19,10 +19,36 @@
...
+    virtual void fill(void *ptr, size_t size, const void *pattern,
+                      size_t pattern_size) = 0;
+    virtual void copy(void *dst_ptr, void *src_ptr, size_t size) = 0;
+};
+
+class HostMemoryAccessor : public MemoryAccessor {
+  public:
+    void fill(void *ptr, size_t size, const void *pattern,
+              size_t pattern_size) override {
+        while (size) {
```
[the remark this answers is not in the document]
Comment (vinser52) [reply test/ipcFixtures.hpp:36]: Done.
Added `assert`

**Pull request merged**